### PR TITLE
Problem: extend ledger wallet address limit to maximum (fix #422)

### LINKED ISSUE #422: Problem: extend ledger wallet address limit to maximum
ledger mode should be expert for this

### PR DESCRIPTION
**Pull request by leejw51crypto**
Solution: extend ledger maximum address index to
99999,9999   <- 9 digits

please change ledger mode to expert

### CODE REVIEW

File: src/config/StaticConfig.ts [hunk 1]
```diff
@@ -5,6 +5,8 @@ export const APP_DB_NAMESPACE = 'data-store';
 export const MARKET_API_BASE_URL = 'https://crypto.org/api';
 export const DEFAULT_CLIENT_MEMO = 'client:chain-desktop-app';
 
+// maximum in ledger: 2147483647
+export const LedgerWalletMaximum = 999999999;
```
Comment (calvinaco) [inline src/config/StaticConfig.ts:9]: Why don't put `2147483647` but `999999999` here?

Comment (leejw51crypto) [reply src/config/StaticConfig.ts:9]: i'll change to 2147483647, ax maximum (inclusive)

**Review by ytbigvinny27 — APPROVED** [review]

**Comment by leejw51crypto** [comment]
chain-maind keys add ok --index=2147483647

from 2147483648, it's error
```
 chain-maind keys add ok --index=2147483648
override the existing name ok [y/N]: y
Error: invalid BIP 32 path m/44'/394'/0'/0/2147483648: strconv.ParseUint: parsing "2147483648": value out of range
Usage:
```

**Comment by crypto-matto** [comment]
Expert Mode checking & error message is added in Sending. 
`Delegate`, `Redelegate`, `Undelegate`, `Withdraw Reward` to be added.

**Comment by eddycpc** [comment]
@leejw51crypto @crypto-matto is this PR ready for merging now?

Also, there are some merge conflicts.

**Comment by crypto-matto** [comment]
> @leejw51crypto @crypto-matto is this PR ready for merging now?
> 
> Also, there are some merge conflicts.

Not yet for my side. I'll update you when it's ready.

**Comment by leejw51crypto** [comment]
checking rebased code now

File: src/pages/send/send.tsx [hunk 2]
```diff
@@ -104,6 +105,19 @@ const FormSend = () => {
...
+        try {
+          const addressprefix = config.network.addressPrefix;
+          const device = createLedgerDevice();
+          const address = await device.getAddress(LedgerWalletMaximum, addressprefix, false);
+          if (address) {
+            setLedgerIsExpertMode(true);
+          }
+        } catch (err) {
+          setLedgerIsExpertMode(false);
+        }
```
Comment (eddycpc) [inline src/pages/send/send.tsx:118]: Can we extract this repeated logic into a function that returns a boolean to set up the `setLedgerIsExpertMode` state ?

**Review by eddycpc — CHANGES_REQUESTED** [review]

File: src/pages/send/send.tsx (hunk 2 above)
Comment (leejw51crypto) [reply src/pages/send/send.tsx:118]: ok, i'll change it

**Review by eddycpc — DISMISSED** [review]
LGTM

**Comment by crypto-matto** [comment]
There are some logics to be fixed. 
There are 3 scenarios:
1. Ledger `connection successful` with `Expert Mode`
2. Ledger `connection successful` with `Normal Mode`
3. Ledger `connection NOT successful`
```
export async function detectLedgerExpertMode(): Promise<boolean> {
  try {
    const device = createLedgerDevice();
    await device.getPubKey(LedgerWalletMaximum, false);
    return true;
  } catch (err) {
    return false;
  }
}
```
For this function checking, it can only differentiate scenario 1 (by returning `true`) from 2 & 3, but cannot differentiate between scenario 2 and 3. Both scenario 2 & 3 will return false. 

My original checking for `isLedgerExpertMode` will return `true` & the error message shows `Please ensure that your have enabled Expert mode on your ledger device.` for scenario 2 only. 

Right now, the error message will show up for scenario 1:
```
      {ledgerIsExpertMode ? (
        <div>Please ensure that your have enabled Expert mode on your ledger device.</div>
      ) : (
        ''
      )}
```

**Comment by leejw51crypto** [comment]
ok, i'll add two error states
1. not connected <- error state
2. need expert mode to generate over 100 address index  <- error state
3. else <- normal , not showing error messag

**Comment by leejw51crypto** [comment]
image

**Comment by leejw51crypto** [comment]
message displayed correctly

File: src/service/LedgerService.ts [hunk 3]
```diff
@@ -16,3 +17,14 @@ export function createLedgerDevice(): ISignerProvider {
...
+
+// true: ledger is expert-mode
+export async function detectLedgerNormalMode(): Promise<boolean> {
+  try {
+    const device = createLedgerDevice();
+    await device.getPubKey(LedgerWalletMaximum, false);
+    return false;
+  } catch (err) {
+    return true;
+  }
```
Comment (calvinaco) [inline src/service/LedgerService.ts:29]: How do you distinguish it is Ledger disconnected or just advance mode is not enabled here?

Comment (leejw51crypto) [reply src/service/LedgerService.ts:29]: when flow  reached here, we don't know whether ledger disconnected or not expert mode.
by calling this api one more time, we can know whether it's by expert mode issue or not.

Comment (leejw51crypto) [reply src/service/LedgerService.ts:29]: to be more clear, i'll add another feedback value to give information.

**Comment by crypto-matto** [comment]
Able to detect scenarios 1,2,3 now 👍🏻 

螢幕截圖 2021-04-14 下午2 29 43

One more minor thing, is it able to remove duplicated `[Conditions not satisfied]` text error ? 

螢幕截圖 2021-04-14 下午2 29 17

**Review by crypto-matto — APPROVED** [review]

**Review by eddycpc — APPROVED** [review]

**Review by calvinaco — DISMISSED** [review]

**Comment by leejw51crypto** [comment]
yes,it's can be removed 
because it's not used in other places

**Pull request merged**